File: api/submit.py
from http.server import BaseHTTPRequestHandler
import json
import os
from urllib.request import Request, urlopen
from urllib.error import URLError

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        content_length = int(self.headers.get('Content-Length', 0))
        body = self.rfile.read(content_length)
        data = json.loads(body)

        # Validate required fields
        first_name = data.get('firstName', '').strip()
        email = data.get('email', '').strip()
        phone = data.get('phone', '').strip()

        if not first_name or not email or not phone:
            self.send_response(400)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({'error': 'All fields are required'}).encode())
            return

        # Submit to Jtek/GHL API
        api_key = os.environ.get('JTEK_API_KEY', '')
        location_id = os.environ.get('JTEK_LOCATION_ID', '')

        contact_data = json.dumps({
            'firstName': first_name,
            'email': email,
            'phone': phone,
            'locationId': location_id,
            'source': '5 Mistakes Guide Lead Magnet',
            'tags': ['lead-magnet', '5-mistakes-guide']
        }).encode()

        req = Request(
            'https://services.leadconnectorhq.com/contacts/',
            data=contact_data,
            headers={
                'Authorization': f'Bearer {api_key}',
                'Content-Type': 'application/json',
                'Version': '2021-04-15',
                'User-Agent': 'JesseTek-LeadMagnet/1.0'
            },
            method='POST'
        )

        try:
            response = urlopen(req)
            result = json.loads(response.read())
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({'success': True, 'contactId': result.get('contact', {}).get('id', '')}).encode())
        except URLError as e:
            error_body = e.read().decode() if hasattr(e, 'read') else str(e)
            # Handle duplicate contact - still a success
            if 'duplicated' in error_body.lower() or 'duplicate' in error_body.lower():
                try:
                    err_data = json.loads(error_body)
                    contact_id = err_data.get('meta', {}).get('contactId', '')
                    # Add tags to existing contact via upsert
                    upsert_data = json.dumps({
                        'firstName': first_name,
                        'email': email,
                        'phone': phone,
                        'locationId': location_id,
                        'source': '5 Mistakes Guide Lead Magnet',
                        'tags': ['lead-magnet', '5-mistakes-guide']
                    }).encode()
                    upsert_req = Request(
                        'https://services.leadconnectorhq.com/contacts/upsert',
                        data=upsert_data,
                        headers={
                            'Authorization': f'Bearer {api_key}',
                            'Content-Type': 'application/json',
                            'Version': '2021-04-15',
                            'User-Agent': 'JesseTek-LeadMagnet/1.0'
                        },
                        method='POST'
                    )
                    upsert_resp = urlopen(upsert_req)
                    upsert_result = json.loads(upsert_resp.read())
                    self.send_response(200)
                    self.send_header('Content-Type', 'application/json')
                    self.send_header('Access-Control-Allow-Origin', '*')
                    self.end_headers()
                    self.wfile.write(json.dumps({'success': True, 'contactId': upsert_result.get('contact', {}).get('id', contact_id)}).encode())
                    return
                except:
                    pass
                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'success': True, 'contactId': '', 'note': 'existing contact'}).encode())
            else:
                self.send_response(500)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Failed to create contact', 'details': error_body}).encode())

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

File: api/submit.js
export default async function handler(req, res) {
  if (req.method === 'OPTIONS') {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
    return res.status(200).end();
  }

  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  res.setHeader('Access-Control-Allow-Origin', '*');

  const { firstName, email, phone, website, elapsed, source, quizScore, quizBand, quizAnswers } = req.body;

  // Honeypot: real users never see this field, so any value means a bot.
  // Time-trap: humans take longer than 2s to fill the form.
  // Silently return 200 so bots think they succeeded and stop retrying.
  if (website || typeof elapsed !== 'number' || elapsed < 2000) {
    return res.status(200).json({ success: true });
  }

  // firstName + email are always required. Phone is required for the guide form
  // (sent as a non-empty string) but optional for the buyer quiz (email-only capture).
  if (!firstName || !email) {
    return res.status(400).json({ error: 'firstName and email are required' });
  }

  const isBuyerQuiz = source === 'buyer-quiz';
  if (!isBuyerQuiz && !phone) {
    return res.status(400).json({ error: 'phone is required' });
  }

  // Quiz bands look like "READY · 80+" — slugify cleanly so GHL workflows can
  // filter by tag name (ready, strong-start, building, early, exploring).
  function slugifyBand(band) {
    if (!band) return 'unknown';
    return String(band)
      .split(/[·•|–\-]/)[0]      // drop the score range after the divider
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, '-') // anything non-alphanumeric → dash
      .replace(/^-+|-+$/g, '');    // trim leading/trailing dashes
  }

  const apiKey = process.env.JTEK_API_KEY || '';
  const locationId = process.env.JTEK_LOCATION_ID || '';

  const contactData = {
    firstName,
    email,
    phone: phone || undefined,
    locationId,
    source: isBuyerQuiz ? 'Buyer Readiness Quiz' : '5 Mistakes Guide Lead Magnet',
    tags: isBuyerQuiz
      ? ['buyer-quiz', `quiz-band-${slugifyBand(quizBand)}`]
      : ['lead-magnet', '5-mistakes-guide'],
  };

  const ghlHeaders = {
    Authorization: `Bearer ${apiKey}`,
    'Content-Type': 'application/json',
    Version: '2021-04-15',
  };

  // Buyer-quiz contacts get a note with their score and answers so Jesse can
  // see context inside the contact (and workflows can email the right plan).
  let notes = '';
  if (isBuyerQuiz) {
    const lines = [`Buyer Readiness Quiz result`];
    if (quizScore !== undefined && quizScore !== null) lines.push(`Score: ${quizScore}`);
    if (quizBand) lines.push(`Band: ${quizBand}`);
    if (quizAnswers && typeof quizAnswers === 'object') {
      lines.push('Answers:');
      Object.entries(quizAnswers).forEach(([q, a]) => lines.push(`  ${q}: ${a}`));
    }
    notes = lines.join('\n');
  }

  async function attachNote(contactId) {
    if (!contactId || !notes) return;
    try {
      await fetch(`https://services.leadconnectorhq.com/contacts/${contactId}/notes`, {
        method: 'POST',
        headers: ghlHeaders,
        body: JSON.stringify({ body: notes }),
      });
    } catch (_) {
      // Best-effort — contact already exists, workflows still fire on tags.
    }
  }

  try {
    const response = await fetch('https://services.leadconnectorhq.com/contacts/', {
      method: 'POST',
      headers: ghlHeaders,
      body: JSON.stringify(contactData),
    });

    if (response.ok) {
      const result = await response.json();
      const contactId = result?.contact?.id || '';
      await attachNote(contactId);
      return res.status(200).json({ success: true, contactId });
    }

    const errorBody = await response.text();
    const isDuplicate = errorBody.toLowerCase().includes('duplicate') || errorBody.toLowerCase().includes('duplicated');

    if (isDuplicate) {
      // Upsert existing contact
      const upsertResponse = await fetch('https://services.leadconnectorhq.com/contacts/upsert', {
        method: 'POST',
        headers: ghlHeaders,
        body: JSON.stringify(contactData),
      });
      const upsertResult = await upsertResponse.json();
      const contactId = upsertResult?.contact?.id || '';
      await attachNote(contactId);
      return res.status(200).json({ success: true, contactId });
    }

    return res.status(500).json({ error: 'Failed to create contact', details: errorBody });
  } catch (err) {
    return res.status(500).json({ error: 'Server error', details: err.message });
  }
}
